### PR TITLE
Quick hack to get datetime working

### PR DESCRIPTION
**Pull request by crisward**
Not sure about the code quality, but this makes date time return a time type in crystal.
If you think it's ok, send me some feedback and I'll chuck some tests in there, and perhaps extend it to cover the other date types.

Thanks.

### CODE REVIEW

**Comment by bcardiff** [comment]
That is probably works in your query only, if the date is followed by other field I doubt it will work.

What is needed is to implement  the `ProtocolBinary::MYSQL_TYPE_DATE, ProtocolBinary::MYSQL_TYPE_DATETIME, ProtocolBinary::MYSQL_TYPE_TIMESTAMP` section of https://dev.mysql.com/doc/internals/en/binary-protocol-value.html

Something along the following lines should work (or arrange the code different to avoid duplication...)

```crystal
def self.read(packet)
  case packet.read_byte!
  when 0
    # build a Time with 0
  when 4
    year = packet.read_fixed_int(2).to_i16
    month = packet.read_byte!
    day = packet.read_byte!
    # build Time
  when 7
    year = packet.read_fixed_int(2).to_i16
    month = packet.read_byte!
    day = packet.read_byte!
    hour = packet.read_byte!
    minute = packet.read_byte!
    second = packet.read_byte!
    # build Time
  when 11
    # the above and ...
    micro_second = packet.read_int
    # build Time
  else
    # protocol error
  end
end
```

PR are definitely welcome!

**Comment by crisward** [comment]
Thanks I'll have a play with this this evening. Hoping the feedback will help hone my crystal code. Can only get better.

**Comment by crisward** [comment]
I've added this

https://github.com/crisward/crystal-mysql/blob/feature/datetime/src/mysql/types.cr#L112

with these tests

https://github.com/crisward/crystal-mysql/blob/feature/datetime/spec/driver_spec.cr#L201

I just need to add the writer now.

1. How to I test the writer? 
2. Does it just need to do the opposite, ie, take a time object and write it out as bytes?

**Comment by tophe** [comment]
hi, it seems there is also a problem reading with  read(String)  when reading a NULL value.

File: src/mysql/types.cr [hunk 1]
```diff
@@ -104,7 +104,17 @@ abstract struct MySql::Type
   decl_type Int24, 0x09u8
   decl_type Date, 0x0au8
   decl_type Time, 0x0bu8
-  decl_type DateTime, 0x0cu8
+  decl_type DateTime, 0x0cu8, ::Bytes do
```
Comment (bcardiff) [inline src/mysql/types.cr]: I just noticed `::Bytes` should be replaced by `::Time` here. check the `  decl_type String, 0xfeu8, ::String do...`

**Comment by bcardiff** [comment]
@crisward 

> How to I test the writer?

There [are test](https://github.com/crisward/crystal-mysql/blob/feature/datetime/spec/driver_spec.cr#L150) that inserts the value in a table using `?` as argument. That will use positional arguments of a prepared statement and use the writers.

For some datatypes it might work to just perform a `SELECT ?` but it since the mapping between crystal types and mysql types are not 1-1 it might not end using the writer you expect. Even so, that is another good test to add [like here](https://github.com/crisward/crystal-mysql/blob/feature/datetime/spec/driver_spec.cr#L116). But mysql might get in the middle and project Date literal as DateTime (might, I am not sure). 

> Does it just need to do the opposite, ie, take a time object and write it out as bytes?

Yes

**Comment by crisward** [comment]
Pretty much what I guessed. I'll let you know when I have a pull request ready. I'll also add a test for null dates in mysql as suggested above by @tophe
